Elevation fix

mschreiber opened the pull request:
Handle the case where the alt is missing in the retrieved data by only calculating the elevationChange if the alt from the last point and the alt from the current point are available

Closes #132

mschreiber committed:
Fix Elevation gain/loss

mschreiber committed:
Fix Elevation gain/loss

mschreiber committed:
Merge branch 'elevationFix' of https://github.com/mschreiber/owntracks-frontend into evaluationFix

linusg commented on src/store/actions.js:
```diff
@@ -130,16 +130,21 @@ const _getTravelStats = (locationHistory) => {
-          const elevationChange = latLng.alt - lastLatLng.alt;
+          // calculate the elevationChange only if there is an alt available
+          let elevationChange = 0;
+          if (lastAlt != null && latLng.alt != null) {
```
```suggestion
          if (lastAlt !== null && latLng.alt !== null) {
```

linusg commented on src/store/actions.js:
```diff
@@ -153,6 +158,9 @@ const _getTravelStats = (locationHistory) => {
             else elevationLoss += -elevationChange;
           }
         }
+        if (latLng.alt != null) {
```
```suggestion
        if (latLng.alt !== null) {
```

linusg reviewed (commented):
Thanks! Two nitpicks

mschreiber committed:
Update src/store/actions.js

Co-authored-by: Linus Groh <mail@linusgroh.de>

mschreiber committed:
Update src/store/actions.js

Co-authored-by: Linus Groh <mail@linusgroh.de>

linusg commented on src/store/actions.js:
```diff
@@ -130,16 +130,21 @@ const _getTravelStats = (locationHistory) => {
-          const elevationChange = latLng.alt - lastLatLng.alt;
+          // calculate the elevationChange only if there is an alt available
+          let elevationChange = 0;
+          if (lastAlt !== null && latLng.alt !== null) {
```
I can find no evidence that an `undefined` alt normalizes to `null` so I think this wouldn't work. Did you test it?

https://github.com/Leaflet/Leaflet/blob/86a910e46eb6712d8e70396115b34ac400869e6f/src/geo/LatLng.js#L46-L48

mschreiber replied:
I am so sorry, you are right, !== null just covers the null case and not undefined. So either just != null or checking both.....

linusg replied:
In what case is `latLng.alt` null or `lastAlt` undefined?

mschreiber replied:
Sorry for all the "noise", your totally right, the null check for latLng.alt and the undefined check for lastAlt are not needed. 
I wrote a small vitest for the _getTravelStats function, should I commit that as well? If so, how should I export the function in order to be able to call it in the test?

linusg replied:
That sounds useful, yeah :)

I think you can just put an `export` before the function and maybe add a comment that it's exported for testing purposes only.

linusg replied:
Looks great, thanks!

mschreiber committed:
Update actions.js

I am so sorry, you are right, !== null just covers the null case and not undefined. So either just != null or checking both.....

mschreiber committed:
removing unnecessary null/undefined checks

mschreiber committed:
adding unittest

mschreiber committed:
adding tests for elevation and distance calculatioadding tests for
elevation and distance calculationn
